configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for INX-MQTT.
keywords:
- IOTA Node 
- Hornet Node
- Dashboard
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-MQTT uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-mqtt` executable.

For example:
```bash
inx-mqtt -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-mqtt -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name        | Description                                                                                                    | Type   | Default value |
| ----------- | -------------------------------------------------------------------------------------------------------------- | ------ | ------------- |
| name        | The optional name of the logger instance. All log messages are prefixed with that name.                        | string | ""            |
| level       | The minimum enabled logging level                                                                              | string | "info"        |
| timeFormat  | Sets the logger's timestamp format. (options: "rfc3339", "rfc3339nano", "datetime", "timeonly", and "iso8601") | string | "rfc3339"     |
| outputPaths | A list of file paths or stdout/stderr to write logging output to                                               | array  | stdout        |

Example:

```json
  {
    "logger": {
      "name": "",
      "level": "info",
      "timeFormat": "rfc3339",
      "outputPaths": [
        "stdout"
      ]
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="mqtt"></a> 4. MQTT

| Name                                 | Description                                                                                             | Type   | Default value                                                                                              |
| ------------------------------------ | ------------------------------------------------------------------------------------------------------- | ------ | ---------------------------------------------------------------------------------------------------------- |
| [websocket](#mqtt_websocket)         | Configuration for websocket                                                                             | object |                                                                                                            |
| [tcp](#mqtt_tcp)                     | Configuration for TCP                                                                                   | object |                                                                                                            |
| [auth](#mqtt_auth)                   | Configuration for auth                                                                                  | object |                                                                                                            |
| publicTopics                         | The MQTT topics which can be subscribed to without authorization. Wildcards using \* are allowed         | array  | commitments/\*<br/>blocks\*<br/>transactions/\*<br/>block-metadata/\*<br/>transaction-metadata/\*<br/>outputs/\* |
| protectedTopics                      | The MQTT topics which only can be subscribed to with valid authorization. Wildcards using \* are allowed | array  |                                                                                                            |
| [subscriptions](#mqtt_subscriptions) | Configuration for subscriptions                                                                         | object |                                                                                                            |
| maximumClientWritesPending           | The maximum number of pending message writes for a client                                               | int    | 8192                                                                                                       |
| clientWriteBufferSize                | The size of the client write buffer                                                                     | int    | 2048                                                                                                       |
| clientReadBufferSize                 | The size of the client read buffer                                                                      | int    | 2048                                                                                                       |

### <a id="mqtt_websocket"></a> Websocket

| Name             | Description                                                                                       | Type    | Default value    |
| ---------------- | ------------------------------------------------------------------------------------------------- | ------- | ---------------- |
| enabled          | Whether to enable the websocket connection of the MQTT broker                                     | boolean | true             |
| bindAddress      | The websocket bind address on which the MQTT broker listens on                                    | string  | "localhost:1888" |
| advertiseAddress | The address of the websocket of the MQTT broker which is advertised to the INX Server (optional). | string  | ""               |

### <a id="mqtt_tcp"></a> TCP

| Name                 | Description                                              | Type    | Default value    |
| -------------------- | -------------------------------------------------------- | ------- | ---------------- |
| enabled              | Whether to enable the TCP connection of the MQTT broker  | boolean | false            |
| bindAddress          | The TCP bind address on which the MQTT broker listens on | string  | "localhost:1883" |
| [tls](#mqtt_tcp_tls) | Configuration for TLS                                    | object  |                  |

### <a id="mqtt_tcp_tls"></a> TLS

| Name            | Description                                                              | Type    | Default value     |
| --------------- | ------------------------------------------------------------------------ | ------- | ----------------- |
| enabled         | Whether to enable TLS for TCP connections                                | boolean | false             |
| privateKeyPath  | The path to the private key file (x509 PEM) for TCP connections with TLS | string  | "private_key.pem" |
| certificatePath | The path to the certificate file (x509 PEM) for TCP connections with TLS | string  | "certificate.pem" |

### <a id="mqtt_auth"></a> Auth

| Name         | Description                                                         | Type   | Default value                                                      |
| ------------ | ------------------------------------------------------------------- | ------ | ------------------------------------------------------------------ |
| passwordSalt | The auth salt used for hashing the passwords of the users           | string | "0000000000000000000000000000000000000000000000000000000000000000" |
| users        | The list of allowed users with their password+salt as a scrypt hash | object | []                                                                 |

### <a id="mqtt_subscriptions"></a> Subscriptions

| Name                           | Description                                                                                                    | Type  | Default value |
| ------------------------------ | -------------------------------------------------------------------------------------------------------------- | ----- | ------------- |
| maxTopicSubscriptionsPerClient | The maximum number of topic subscriptions per client before the client gets dropped (DOS protection)           | int   | 1000          |
| topicsCleanupThresholdCount    | The number of deleted topics that trigger a garbage collection of the subscription manager                     | int   | 10000         |
| topicsCleanupThresholdRatio    | The ratio of subscribed topics to deleted topics that trigger a garbage collection of the subscription manager | float | 1.0           |

Example:

```json
  {
    "mqtt": {
      "websocket": {
        "enabled": true,
        "bindAddress": "localhost:1888",
        "advertiseAddress": ""
      },
      "tcp": {
        "enabled": false,
        "bindAddress": "localhost:1883",
        "tls": {
          "enabled": false,
          "privateKeyPath": "private_key.pem",
          "certificatePath": "certificate.pem"
        }
      },
      "auth": {
        "passwordSalt": "0000000000000000000000000000000000000000000000000000000000000000",
        "users": null
      },
      "publicTopics": [
        "commitments/*",
        "blocks*",
        "transactions/*",
        "block-metadata/*",
        "transaction-metadata/*",
        "outputs/*"
      ],
      "protectedTopics": [],
      "subscriptions": {
        "maxTopicSubscriptionsPerClient": 1000,
        "topicsCleanupThresholdCount": 10000,
        "topicsCleanupThresholdRatio": 1
      },
      "maximumClientWritesPending": 8192,
      "clientWriteBufferSize": 2048,
      "clientReadBufferSize": 2048
    }
  }
```

## <a id="profiling"></a> 5. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling component is enabled        | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="prometheus"></a> 6. Prometheus

| Name            | Description                                                     | Type    | Default value    |
| --------------- | --------------------------------------------------------------- | ------- | ---------------- |
| enabled         | Whether the prometheus plugin is enabled                        | boolean | false            |
| bindAddress     | The bind address on which the Prometheus HTTP server listens on | string  | "localhost:9312" |
| mqttMetrics     | Whether to include MQTT metrics                                 | boolean | true             |
| goMetrics       | Whether to include go metrics                                   | boolean | false            |
| processMetrics  | Whether to include process metrics                              | boolean | false            |
| promhttpMetrics | Whether to include promhttp metrics                             | boolean | false            |

Example:

```json
  {
    "prometheus": {
      "enabled": false,
      "bindAddress": "localhost:9312",
      "mqttMetrics": true,
      "goMetrics": false,
      "processMetrics": false,
      "promhttpMetrics": false
    }
  }
```



documentation/docs/configuration.md
<!---
!!! DO NOT MODIFY !!!

This file is auto-generated by the gendoc tool based on the source code of the app.
-->
---
description: This section describes the configuration parameters and their types for INX-MQTT.
keywords:
- IOTA Node 
- Hornet Node
- Dashboard
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-MQTT uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-mqtt` executable.

For example:
```bash
inx-mqtt -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-mqtt -h --full
```

## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name              | Description                                                                 | Type    | Default value |
| ----------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level             | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller     | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel   | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding          | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| outputPaths       | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents     | Prevents log messages from being raced as events                            | boolean | true          |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30
    }
  }
```

## <a id="mqtt"></a> 4. MQTT

| Name                                 | Description                                   | Type   | Default value |
| ------------------------------------ | --------------------------------------------- | ------ | ------------- |
| bufferSize                           | The size of the client buffers in bytes       | int    | 0             |
| bufferBlockSize                      | The size per client buffer R/W block in bytes | int    | 0             |
| [subscriptions](#mqtt_subscriptions) | Configuration for subscriptions               | object |               |
| [websocket](#mqtt_websocket)         | Configuration for websocket                   | object |               |
| [tcp](#mqtt_tcp)                     | Configuration for TCP                         | object |               |

### <a id="mqtt_subscriptions"></a> Subscriptions

| Name                           | Description                                                                                                    | Type  | Default value |
| ------------------------------ | -------------------------------------------------------------------------------------------------------------- | ----- | ------------- |
| maxTopicSubscriptionsPerClient | The maximum number of topic subscriptions per client before the client gets dropped (DOS protection)           | int   | 1000          |
| topicsCleanupThresholdCount    | The number of deleted topics that trigger a garbage collection of the subscription manager                     | int   | 10000         |
| topicsCleanupThresholdRatio    | The ratio of subscribed topics to deleted topics that trigger a garbage collection of the subscription manager | float | 1.0           |

### <a id="mqtt_websocket"></a> Websocket

| Name             | Description                                                                                       | Type    | Default value    |
| ---------------- | ------------------------------------------------------------------------------------------------- | ------- | ---------------- |
| enabled          | Whether to enable the websocket connection of the MQTT broker                                     | boolean | true             |
| bindAddress      | The websocket bind address on which the MQTT broker listens on                                    | string  | "localhost:1888" |
| advertiseAddress | The address of the websocket of the MQTT broker which is advertised to the INX Server (optional). | string  | ""               |

### <a id="mqtt_tcp"></a> TCP

| Name                   | Description                                              | Type    | Default value    |
| ---------------------- | -------------------------------------------------------- | ------- | ---------------- |
| enabled                | Whether to enable the TCP connection of the MQTT broker  | boolean | false            |
| bindAddress            | The TCP bind address on which the MQTT broker listens on | string  | "localhost:1883" |
| [auth](#mqtt_tcp_auth) | Configuration for auth                                   | object  |                  |
| [tls](#mqtt_tcp_tls)   | Configuration for TLS                                    | object  |                  |

### <a id="mqtt_tcp_auth"></a> Auth

| Name         | Description                                                         | Type    | Default value                                                      |
| ------------ | ------------------------------------------------------------------- | ------- | ------------------------------------------------------------------ |
| enabled      | Whether to enable auth for TCP connections                          | boolean | false                                                              |
| passwordSalt | The auth salt used for hashing the passwords of the users           | string  | "0000000000000000000000000000000000000000000000000000000000000000" |
| users        | The list of allowed users with their password+salt as a scrypt hash | object  | []                                                                 |

### <a id="mqtt_tcp_tls"></a> TLS

| Name            | Description                                                              | Type    | Default value     |
| --------------- | ------------------------------------------------------------------------ | ------- | ----------------- |
| enabled         | Whether to enable TLS for TCP connections                                | boolean | false             |
| privateKeyPath  | The path to the private key file (x509 PEM) for TCP connections with TLS | string  | "private_key.pem" |
| certificatePath | The path to the certificate file (x509 PEM) for TCP connections with TLS | string  | "certificate.pem" |

Example:

```json
  {
    "mqtt": {
      "bufferSize": 0,
      "bufferBlockSize": 0,
      "subscriptions": {
        "maxTopicSubscriptionsPerClient": 1000,
        "topicsCleanupThresholdCount": 10000,
        "topicsCleanupThresholdRatio": 1
      },
      "websocket": {
        "enabled": true,
        "bindAddress": "localhost:1888",
        "advertiseAddress": ""
      },
      "tcp": {
        "enabled": false,
        "bindAddress": "localhost:1883",
        "auth": {
          "enabled": false,
          "passwordSalt": "0000000000000000000000000000000000000000000000000000000000000000",
          "users": null
        },
        "tls": {
          "enabled": false,
          "privateKeyPath": "private_key.pem",
          "certificatePath": "certificate.pem"
        }
      }
    }
  }
```

## <a id="profiling"></a> 5. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling plugin is enabled           | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="prometheus"></a> 6. Prometheus

| Name            | Description                                                     | Type    | Default value    |
| --------------- | --------------------------------------------------------------- | ------- | ---------------- |
| enabled         | Whether the prometheus plugin is enabled                        | boolean | false            |
| bindAddress     | The bind address on which the Prometheus HTTP server listens on | string  | "localhost:9312" |
| mqttMetrics     | Whether to include MQTT metrics                                 | boolean | true             |
| goMetrics       | Whether to include go metrics                                   | boolean | false            |
| processMetrics  | Whether to include process metrics                              | boolean | false            |
| promhttpMetrics | Whether to include promhttp metrics                             | boolean | false            |

Example:

```json
  {
    "prometheus": {
      "enabled": false,
      "bindAddress": "localhost:9312",
      "mqttMetrics": true,
      "goMetrics": false,
      "processMetrics": false,
      "promhttpMetrics": false
    }
  }
```



components/mqtt/params.go
package mqtt

import "github.com/iotaledger/hive.go/app"

type ParametersMQTT struct {
	Websocket struct {
		Enabled          bool   `default:"true" usage:"whether to enable the websocket connection of the MQTT broker"`
		BindAddress      string `default:"localhost:1888" usage:"the websocket bind address on which the MQTT broker listens on"`
		AdvertiseAddress string `default:"" usage:"the address of the websocket of the MQTT broker which is advertised to the INX Server (optional)."`
	}

	TCP struct {
		Enabled     bool   `default:"false" usage:"whether to enable the TCP connection of the MQTT broker"`
		BindAddress string `default:"localhost:1883" usage:"the TCP bind address on which the MQTT broker listens on"`

		TLS struct {
			Enabled         bool   `default:"false" usage:"whether to enable TLS for TCP connections"`
			PrivateKeyPath  string `default:"private_key.pem" usage:"the path to the private key file (x509 PEM) for TCP connections with TLS"`
			CertificatePath string `default:"certificate.pem" usage:"the path to the certificate file (x509 PEM) for TCP connections with TLS"`
		} `name:"tls"`
	} `name:"tcp"`

	Auth struct {
		PasswordSalt string            `default:"0000000000000000000000000000000000000000000000000000000000000000" usage:"the auth salt used for hashing the passwords of the users"`
		Users        map[string]string `usage:"the list of allowed users with their password+salt as a scrypt hash"`
	}

	PublicTopics    []string `usage:"the MQTT topics which can be subscribed to without authorization. Wildcards using * are allowed"`
	ProtectedTopics []string `usage:"the MQTT topics which only can be subscribed to with valid authorization. Wildcards using * are allowed"`

	Subscriptions struct {
		MaxTopicSubscriptionsPerClient int     `default:"1000" usage:"the maximum number of topic subscriptions per client before the client gets dropped (DOS protection)"`
		TopicsCleanupThresholdCount    int     `default:"10000" usage:"the number of deleted topics that trigger a garbage collection of the subscription manager"`
		TopicsCleanupThresholdRatio    float32 `default:"1.0" usage:"the ratio of subscribed topics to deleted topics that trigger a garbage collection of the subscription manager"`
	}

	MaximumClientWritesPending int `default:"8192" usage:"the maximum number of pending message writes for a client"`
	ClientWriteBufferSize      int `default:"2048" usage:"the size of the client write buffer"`
	ClientReadBufferSize       int `default:"2048" usage:"the size of the client read buffer"`
}

var ParamsMQTT = &ParametersMQTT{
	PublicTopics: []string{
		"commitments/*",
		"blocks*",
		"transactions/*",
		"block-metadata/*",
		"transaction-metadata/*",
		"outputs/*",
	},
	ProtectedTopics: []string{},
}

var params = &app.ComponentParams{
	Params: map[string]any{
		"mqtt": ParamsMQTT,
	},
	Masked: []string{"mqtt.auth.passwordSalt"},
}
